sdk/core/core/src/az_url.c

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "az_hex_private.h"
+#include <az_precondition_internal.h>
+#include <az_url_internal.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD AZ_INLINE bool _az_url_should_encode(uint8_t c)
+{
+  switch (c)
+  {
+    case '-':
+    case '_':
+    case '.':
+    case '~':
+      return false;
+    default:
+      return !(('0' <= c && c <= '9') || ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z'));
+  }
+}
+
+AZ_NODISCARD az_result _az_url_encode(az_span destination, az_span source, int32_t* out_length)
+{
+  AZ_PRECONDITION_NOT_NULL(out_length);
+
+  int32_t const input_size = az_span_size(source);
+
+  int32_t result_size = 0;
+  for (int32_t i = 0; i < input_size; ++i)
+  {
+    result_size += _az_url_should_encode(az_span_ptr(source)[i]) ? 3 : 1;
+  }
+
+  if (az_span_size(destination) < result_size)
+  {
+    return AZ_ERROR_INSUFFICIENT_SPAN_SIZE;
+  }
+
+  uint8_t* p_s = az_span_ptr(source);
```

Comment:
NitL p_s, p_d, s are not descriptive.  Its very easy to mix them up

Reply:
Given this is primarily an implementation move, I would prefer any such changes be made in a separate PR so it can be reviewed independantly.

Reply:
pointer to source, pointer to destination... I find it ok, but it is subjective. I can change it, I think I am fine with more descriptive names. Should I go with src_ptr, and dest_ptr?

Reply:
I like `src_ptr`, `dest_ptr`.

Reply:
Done.